src/components/views/rooms/MessageComposerInput.js

```diff
@@ -1204,50 +1204,44 @@ export default class MessageComposerInput extends React.Component {
     };
 
     onVerticalArrow = (e, up) => {
-        if (e.ctrlKey || e.altKey || e.metaKey) {
-            return;
-        }
-
-        // Select history only if we are not currently auto-completing
-        if (this.autocomplete.state.completionList.length === 0) {
-            const selection = this.state.editorState.selection;
-
-            // selection must be collapsed
-            if (!selection.isCollapsed) return;
-            const document = this.state.editorState.document;
-
-            // and we must be at the edge of the document (up=start, down=end)
-            if (up) {
-                if (!selection.anchor.isAtStartOfNode(document)) return;
-
-                if (!e.shiftKey) {
-                    const editEvent = findEditableEvent(this.props.room, false);
-                    if (editEvent) {
-                        // We're selecting history, so prevent the key event from doing anything else
-                        e.preventDefault();
-                        dis.dispatch({
-                            action: 'edit_event',
-                            event: editEvent,
-                        });
-                    }
-                    return;
+        if (e.ctrlKey || e.shiftKey || e.metaKey) return;
+
+        // Select history
+        const selection = this.state.editorState.selection;
+
+        // selection must be collapsed
+        if (!selection.isCollapsed) return;
+        const document = this.state.editorState.document;
+
+        // and we must be at the edge of the document (up=start, down=end)
+        if (up) {
+            if (!selection.anchor.isAtStartOfNode(document)) return;
+
+            if (!e.altKey) {
+                const editEvent = findEditableEvent(this.props.room, false);
+                if (editEvent) {
+                    // We're selecting history, so prevent the key event from doing anything else
+                    e.preventDefault();
+                    dis.dispatch({
+                        action: 'edit_event',
+                        event: editEvent,
+                    });
                 }
-            } else {
-                if (!selection.anchor.isAtEndOfNode(document)) return;
+                return;
             }
+        }
 
-            const selected = this.selectHistory(up);
-            if (selected) {
-                // We're selecting history, so prevent the key event from doing anything else
-                e.preventDefault();
-            }
+        const selected = this.selectHistory(up);
```

Comment:
the logic gets a bit "double negative" and scattered in here, in that it's not immediately obvious that this code will only run when alt is pressed. How about:

```
if (e.altKey) {
   // restore history
} else if (up && !(e.ctrlKey || e.metaKey || e.shiftKey) && caretAtStart) {
  // edit last sent event
}
```